### ISSUE TITLE
Figure out best way to refactor models

### ISSUE THREAD

**Issue by philippmwirth** [lightly-ai/lightly#377]
# Figure out best way to refactor models

The current setup makes it hard to implement new and upcoming models. Additionally, there is a lot of code which is repeated between the different implementations. The goal of this issue is to find the best way to refactor our models such that
- there is a centralized concept of the prediction/projection heads
- the code and workflow becomes more transparent to our users
- (as a consequence) it's easier to add new models

@IgorSusmelj can you add the feedback you received from Anna in my list above or as a separate comment? Thanks!

**Comment by philippmwirth** [comment]
**THIS IS A DRAFT**


# Proposal for Refactoring


```
lightly/
L models/
   L backbones/
   L heads/
   L encoders/
```

## Backbones
ResNet, whatever backbones we will implement

## Heads
- Base class
- Configurable prediction and projection heads
- Predefined prediction and projection heads (e.g. `MoCoV2PredictionHead`)

## Encoders
- Consists of backbone + head(s)
- Differentiate between encoder and momentum encoder

## Proposed Workflow
...

**Comment by IgorSusmelj** [comment]
I think there is no need to put heads into a folder as we won't have many. If we have a way to generate heads they don't require lots of code. I'm only aware of different combinations of whether there is a batchnorm layer or how many linear layers there are. I don't see a need for more than one file for them. 

Regarding the encoders. I would not call them encoders. I would maybe make a folder called `modules` for things like momentum encoder, split batch norm.

The models themselves I would leave just in the `model` folder. That makes it easy to find them and gives the whole setup a simple hierarchy. It also follows for example the torchvision models setup.

I would suggest something like this:
```
lightly/
L models/
   L modules/
       L heads.py
       L batch_norm.py
       L momentum_encoder.py
   L barlow_twins.py
   L moco.py
   L simclr.py
   L simsiam.py
   L ...
```

**Comment by philippmwirth** [comment]
I am on board with the structure you proposed. 

I'd still prefer to call the models `encoder` as this is consistent with literature, see e.g. MoCo

![grafik](https://user-images.githubusercontent.com/65946090/119090907-85143500-ba0c-11eb-8405-c4fda032014b.png)

Then we could have a `MoCoEncoder` class in `moco.py` and some functionality to get the `MoCoMomentumEncoder` and then the forward pass through both of them has to be done one after another (no ifs and elses from our side).

**Comment by IgorSusmelj** [comment]
I get your point regarding the naming and the papers. However, if you look at all the actual implementations barely anyone uses `encoder`. (`MoCo`: https://github.com/facebookresearch/moco/blob/master/moco/builder.py, `BarlowTwins`: https://github.com/facebookresearch/barlowtwins/blob/main/main.py, `BYOL`: https://github.com/lucidrains/byol-pytorch/blob/master/byol_pytorch/byol_pytorch.py, ) 
I think the term `encoder` is solely used in the literature. Since we try to build an easy-to-use and compatible framework with others I would stick with the current namings of the models. `MoCo`, `BarlowTwins`, ...

**Comment by MalteEbner** [comment]
I strongly agree with @philippmwirth that we should call the encoder encoder. This term is also used in other overview graphs, as the one from SimSiam. It also has the advantage that we could use these overview graphs or similar versions directly in our documentation.
<img width="786" alt="image" src="https://user-images.githubusercontent.com/20324507/128057374-fc41ee7b-3364-4ff9-af5f-a923ff0ae186.png">

Furthermore I would not think of "Moco", "SimCLR" and so on as a model, but rather a system consisting of an augmenter (some systems/papers use crops of different sizes), an encoder, a prediction/projection head, and a pretext task.
The pretext task itself contains things such as a memory bank, momentum encoder, NN replacer (like for NNCLR), assignment swapper (like for Swav), soft label methods and the loss function. In general, it defines which samples should be considered as similar or different and how embeddings should be pushed together or away.

**Comment by MalteEbner** [comment]
After a discussion, this is closed in favour of #378 and #379
